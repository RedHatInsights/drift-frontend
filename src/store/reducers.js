import React from 'react';
import { mergeArraysByKey } from '@redhat-cloud-services/frontend-components-utilities/files/helpers';
import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import HistoricalProfilesPopover from '../SmartComponents/HistoricalProfilesPopover/HistoricalProfilesPopover';
import SystemKebab from '../SmartComponents/BaselinesPage/EditBaselinePage/SystemNotification/SystemKebab/SystemKebab';

import types from '../SmartComponents/modules/types';
import helpers from '../SmartComponents/helpers';

function selectedReducer(
    INVENTORY_ACTIONS, baselineId, createBaselineModal, historicalProfiles, hasMultiSelect, hasHistoricalDropdown,
    deselectHistoricalProfiles, isAddSystemNotifications, selectHistoricProfiles, systemNotificationIds, selectSystemsToAdd,
    deleteNotifications
) {
    let newColumns;
    let systemIds;

    return applyReducerHash({
        [INVENTORY_ACTIONS.LOAD_ENTITIES_FULFILLED]: (state, action) => {
            newColumns = [ ...state.columns ];

            for (let i = 0; i < action.payload.results.length; i += 1) {
                if (state.selectedSystemIds.includes(action.payload.results[i].id)) {
                    action.payload.results[i].selected = true;
                }
            }

            let rows = mergeArraysByKey([ action.payload.results, state.rows ]);

            if (historicalProfiles !== undefined) {
                /*eslint-disable camelcase*/
                if (!newColumns.find(({ key }) => key === 'historical_profiles') && hasHistoricalDropdown) {
                    newColumns.push({
                        key: 'historical_profiles',
                        title: 'Historical profiles',
                        props: {
                            isStatic: true,
                            width: 10
                        }
                    });
                }

                if (newColumns.find(({ key }) => key === 'system_notification')) {
                    newColumns.splice(newColumns.findIndex(({ key }) => key === 'system_notification'), 1);
                }

                rows.forEach(function(row) {
                    let badgeCount = 0;
                    let systemInfo = {
                        id: row.id,
                        last_updated: row.updated
                    };

                    historicalProfiles.forEach(function(profile) {
                        if (profile.system_id === row.id) {
                            badgeCount += 1;
                        }
                    });

                    row.historical_profiles = <React.Fragment>
                        <HistoricalProfilesPopover
                            system={ systemInfo }
                            hasBadge={ true }
                            badgeCount={ badgeCount }
                            hasMultiSelect={ hasMultiSelect }
                            selectHistoricProfiles={ selectHistoricProfiles }
                        />
                    </React.Fragment>;
                });
            }

            if (isAddSystemNotifications) {
                rows.forEach(function(row) {
                    if (systemNotificationIds.includes(row.id)) {
                        row.selected = true;
                        row.disableSelection = true;
                    }
                });
            }

            if (baselineId && !isAddSystemNotifications) {
                if (!newColumns.find(({ key }) => key === 'system_notification')) {
                    newColumns.push({
                        key: 'system_notification',
                        title: '',
                        props: {
                            isStatic: true
                        }
                    });
                }

                if (newColumns.find(({ key }) => key === 'historical_profiles')) {
                    newColumns.splice(newColumns.findIndex(({ key }) => key === 'historical_profiles'), 1);
                }

                rows.forEach(function(row) {
                    systemIds = [ row.id ];

                    row.system_notification = <React.Fragment>
                        <SystemKebab
                            systemIds={ systemIds }
                            systemName={ row.display_name }
                            deleteNotifications={ deleteNotifications }
                        />
                    </React.Fragment>;
                });
            }
            /*eslint-enable camelcase*/

            /* Hide link on systems table */
            newColumns.forEach(function(column, index) {
                if (!column.props) {
                    column.props = {};
                }

                if (hasHistoricalDropdown) {
                    if (column.key === 'display_name' || column.key === 'display_selected_hsp') {
                        column.props.width = 20;
                        if (newColumns.find(({ key }) => key === 'display_selected_hsp')) {
                            let displaySelectedHSP = newColumns.splice(newColumns.findIndex(({ key }) => key === 'display_selected_hsp'), 1);
                            newColumns.splice(index, 1, displaySelectedHSP[0]);
                        }
                    } else {
                        column.props.width = 10;
                    }
                } else {
                    column.props.width = null;
                }

                if (column.composed) {
                    delete column.composed;
                }
            });

            return {
                ...state,
                columns: newColumns,
                rows: state.selectedHSP
                    ? helpers.buildSystemsTableWithSelectedHSP(rows, state.selectedHSP, deselectHistoricalProfiles)
                    : rows
            };
        },
        [types.DRIFT_CLEAR_ALL_FILTERS]: (state) => ({
            ...state,
            activeFilters: undefined
        }),
        [types.UPDATE_COLUMNS]: (state, action) => {
            let nameColumn = {
                key: action.payload,
                title: 'Name',
                props: { width: 20 }
            };
            newColumns = [ ...state.columns ];
            newColumns.shift();
            newColumns.unshift(nameColumn);

            return {
                ...state,
                columns: newColumns
            };
        },
        [types.SELECT_ENTITY]: (state, action) => {
            let id = action.payload.id;
            let selected = action.payload.selected;
            let { selectedSystemIds } = state;
            let newRows = [];

            if (id === 0) {
                if (createBaselineModal) {
                    newRows = state.rows.map(function (oneRow) {
                        oneRow.selected = false;
                        return oneRow;
                    });

                    selectedSystemIds = [];
                } else {
                    if (selected) {
                        let ids = state.rows.map(function (item) {
                            return item.id;
                        });
                        selectedSystemIds = [ ...new Set(selectedSystemIds.concat(ids)) ];
                    } else {
                        if (action.payload.bulk) {
                            selectedSystemIds = [];
                        } else {
                            for (let i = 0; i < state.rows.length; i += 1) {
                                selectedSystemIds = selectedSystemIds.filter(item => item !== state.rows[i].id);
                            }
                        }
                    }
                }
            } else if (selected && id !== null) {
                if (createBaselineModal) {
                    newRows = state.rows.map(function (oneRow) {
                        if (oneRow.id === id) {
                            oneRow.selected = true;
                        } else {
                            oneRow.selected = false;
                        }

                        return oneRow;
                    });

                    selectedSystemIds.pop();
                }

                selectedSystemIds.push(id);
            } else {
                selectedSystemIds = selectedSystemIds.filter(item => item !== id);
            }

            if (isAddSystemNotifications) {
                selectSystemsToAdd(selectedSystemIds);
            }

            return {
                ...state,
                selectedSystemIds,
                rows: newRows.length > 0 ? newRows : state.rows,
                columns: newColumns
            };
        },
        [types.SET_SELECTED_SYSTEM_IDS]: (state, action) => {
            return {
                ...state,
                selectedSystemIds: action.payload.selectedSystemIds
            };
        },
        [types.SELECT_SINGLE_HSP]: (state, action) => {
            return {
                ...state,
                rows: helpers.buildSystemsTableWithSelectedHSP([ ...state.rows ], action.payload, deselectHistoricalProfiles),
                columns: newColumns,
                selectedSystemIds: [],
                selectedHSP: action.payload
            };
        }
    });
}

export default selectedReducer;
