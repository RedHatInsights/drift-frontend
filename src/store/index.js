import promiseMiddleware from 'redux-promise-middleware';
import { getRegistry } from '@redhat-cloud-services/frontend-components-utilities/files/Registry';

import { compareReducer, globalFilterReducer } from '../SmartComponents/modules/reducers';
import { addSystemModalReducer } from '../SmartComponents/AddSystemModal/redux/addSystemModalReducer';
import { baselinesTableRootReducer } from '../SmartComponents/BaselinesTable/redux';
import { filterDropdownReducer } from '../SmartComponents/DriftPage/FilterDropDown/redux/filterDropdownReducer';
import { createBaselineModalReducer } from '../SmartComponents/BaselinesPage/CreateBaselineModal/redux/reducers';
import { editBaselineReducer } from '../SmartComponents/BaselinesPage/EditBaselinePage/redux/reducers';
import { historicProfilesReducer } from '../SmartComponents/HistoricalProfilesPopover/redux/reducers';
import { systemNotificationsReducer } from '../SmartComponents/BaselinesPage/EditBaselinePage/SystemNotification/redux/reducer';

import MiddlewareListener from '@redhat-cloud-services/frontend-components-utilities/files/MiddlewareListener';
import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
export { default as reducers } from './reducers';

let registry;
let middlewareListener;

export const createMiddlewareListener = () => {
    middlewareListener = new MiddlewareListener();
    return middlewareListener;
};

export function init (...middleware) {
    createMiddlewareListener();

    registry = getRegistry({}, [
        promiseMiddleware(),
        middlewareListener.getMiddleware(),
        ...middleware.filter(item => typeof item !== 'undefined')
    ]);

    registry.register({
        compareState: compareReducer,
        addSystemModalState: addSystemModalReducer,
        baselinesTableState: baselinesTableRootReducer,
        filterDropdownOpened: filterDropdownReducer,
        createBaselineModalState: createBaselineModalReducer,
        editBaselineState: editBaselineReducer,
        historicProfilesState: historicProfilesReducer,
        systemNotificationsState: systemNotificationsReducer,
        notifications,
        globalFilterState: globalFilterReducer
    });

    return registry;
}

export function getStore () {
    return registry.getStore();
}

export function register (...args) {
    return registry.register(...args);
}

export function addNewListener ({ actionType, callback }) {
    return middlewareListener.addNew({
        on: actionType,
        callback
    });
}
