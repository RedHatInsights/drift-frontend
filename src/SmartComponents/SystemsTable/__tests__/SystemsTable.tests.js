import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import toJson from 'enzyme-to-json';

import ConnectedSystemsTable from '../SystemsTable';
import { createMiddlewareListener } from '../../../store';
import fixtures from './fixtures';

const middlewareListener = createMiddlewareListener();
middlewareListener.getMiddleware();

const MockComponent = jest.fn(({ children, loaded }) => {
    return children && loaded ? children : 'Loading...';
});

describe('ConnectedSystemsTable', () => {
    let initialState;
    let mockStore;
    let props;

    beforeEach(() => {
        mockStore = configureStore();
        initialState = {
            selectedSystemIds: [],
            createBaselineModal: false,
            hasHistoricalDropdown: false,
            historicalProfiles: [],
            hasMultiSelect: true,
            entities: {
                columns: fixtures.columns,
                rows: fixtures.rows,
                total: 3,
                count: 3,
                selectedSystemIds: []
            },
            selectVariant: 'checkbox',
            setSelectedSystemIds: jest.fn(),
            driftClearFilters: jest.fn(),
            selectHistoricProfiles: jest.fn(),
            updateColumns: jest.fn(),
            selectEntities: jest.fn()
        };
        props = {
            hasInventoryReadPermissions: true
        };

        global.window.insights = {
            loadInventory: jest.fn(() => {
                return Promise.resolve({
                    inventoryConnector: () => ({
                        InventoryDetails: MockComponent
                    }),
                    INVENTORY_ACTION_TYPES: {},
                    mergeWithEntities: () => ({})
                });
            })
        };
    });

    it('should render correctly', () => {
        const store = mockStore(initialState);

        const wrapper = mount(
            <MemoryRouter keyLength={ 0 }>
                <Provider store={ store }>
                    <ConnectedSystemsTable { ...props } />
                </Provider>
            </MemoryRouter>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it.skip('should render radio correctly', () => {
        initialState.selectVariant = 'radio';
        const store = mockStore(initialState);

        const wrapper = mount(
            <MemoryRouter keyLength={ 0 }>
                <Provider store={ store }>
                    <ConnectedSystemsTable { ...props } />
                </Provider>
            </MemoryRouter>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
