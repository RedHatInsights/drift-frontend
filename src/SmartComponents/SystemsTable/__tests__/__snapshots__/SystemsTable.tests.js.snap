// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedSystemsTable should render correctly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(SystemsTable)
        hasInventoryReadPermissions={true}
      >
        <SystemsTable
          driftClearFilters={[Function]}
          hasInventoryReadPermissions={true}
          selectEntities={[Function]}
          selectHistoricProfiles={[Function]}
          selectedSystemIds={Array []}
          setSelectedSystemIds={[Function]}
          updateColumns={[Function]}
        >
          <ForwardRef
            bulkSelect={
              Object {
                "checked": null,
                "count": 0,
                "isDisabled": true,
                "items": Array [
                  Object {
                    "onClick": [Function],
                    "title": "Select none (0)",
                  },
                  Object {
                    "onClick": [Function],
                    "title": "Select page (0)",
                  },
                ],
                "onSelect": [Function],
              }
            }
            customFilters={
              Object {
                "filter": Object {
                  "system_profile": Object {},
                },
                "tags": undefined,
              }
            }
            fallback={
              <Bullseye>
                <Spinner
                  size="xl"
                />
              </Bullseye>
            }
            getEntities={[Function]}
            noDetail={true}
            onLoad={[Function]}
            showTags={true}
            tableProps={
              Object {
                "canSelectAll": false,
                "ouiaId": "systems-table",
                "selectVariant": undefined,
              }
            }
          >
            <BaseInvTable
              bulkSelect={
                Object {
                  "checked": null,
                  "count": 0,
                  "isDisabled": true,
                  "items": Array [
                    Object {
                      "onClick": [Function],
                      "title": "Select none (0)",
                    },
                    Object {
                      "onClick": [Function],
                      "title": "Select page (0)",
                    },
                  ],
                  "onSelect": [Function],
                }
              }
              customFilters={
                Object {
                  "filter": Object {
                    "system_profile": Object {},
                  },
                  "tags": undefined,
                }
              }
              fallback={
                <Bullseye>
                  <Spinner
                    size="xl"
                  />
                </Bullseye>
              }
              getEntities={[Function]}
              innerRef={null}
              noDetail={true}
              onLoad={[Function]}
              showTags={true}
              tableProps={
                Object {
                  "canSelectAll": false,
                  "ouiaId": "systems-table",
                  "selectVariant": undefined,
                }
              }
            >
              <Suspense
                fallback={
                  <Bullseye>
                    <Spinner
                      size="xl"
                    />
                  </Bullseye>
                }
              >
                <ForwardRef
                  ErrorComponent={
                    <UNDEFINED
                      bulkSelect={
                        Object {
                          "checked": null,
                          "count": 0,
                          "isDisabled": true,
                          "items": Array [
                            Object {
                              "onClick": [Function],
                              "title": "Select none (0)",
                            },
                            Object {
                              "onClick": [Function],
                              "title": "Select page (0)",
                            },
                          ],
                          "onSelect": [Function],
                        }
                      }
                      component="InventoryTable"
                      customFilters={
                        Object {
                          "filter": Object {
                            "system_profile": Object {},
                          },
                          "tags": undefined,
                        }
                      }
                      fallback={
                        <Bullseye>
                          <Spinner
                            size="xl"
                          />
                        </Bullseye>
                      }
                      getEntities={[Function]}
                      innerRef={null}
                      noDetail={true}
                      onLoad={[Function]}
                      showTags={true}
                      tableProps={
                        Object {
                          "canSelectAll": false,
                          "ouiaId": "systems-table",
                          "selectVariant": undefined,
                        }
                      }
                    />
                  }
                  appName="chrome"
                  bulkSelect={
                    Object {
                      "checked": null,
                      "count": 0,
                      "isDisabled": true,
                      "items": Array [
                        Object {
                          "onClick": [Function],
                          "title": "Select none (0)",
                        },
                        Object {
                          "onClick": [Function],
                          "title": "Select page (0)",
                        },
                      ],
                      "onSelect": [Function],
                    }
                  }
                  customFilters={
                    Object {
                      "filter": Object {
                        "system_profile": Object {},
                      },
                      "tags": undefined,
                    }
                  }
                  fallback={
                    <Bullseye>
                      <Spinner
                        size="xl"
                      />
                    </Bullseye>
                  }
                  getEntities={[Function]}
                  history={
                    Object {
                      "action": "POP",
                      "block": [Function],
                      "canGo": [Function],
                      "createHref": [Function],
                      "entries": Array [
                        Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        },
                      ],
                      "go": [Function],
                      "goBack": [Function],
                      "goForward": [Function],
                      "index": 0,
                      "length": 1,
                      "listen": [Function],
                      "location": Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      },
                      "push": [Function],
                      "replace": [Function],
                    }
                  }
                  innerRef={null}
                  module="./InventoryTable"
                  noDetail={true}
                  onLoad={[Function]}
                  scope="chrome"
                  showTags={true}
                  store={
                    Object {
                      "clearActions": [Function],
                      "dispatch": [Function],
                      "getActions": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                    }
                  }
                  tableProps={
                    Object {
                      "canSelectAll": false,
                      "ouiaId": "systems-table",
                      "selectVariant": undefined,
                    }
                  }
                >
                  <BaseScalprumComponent
                    ErrorComponent={
                      <UNDEFINED
                        bulkSelect={
                          Object {
                            "checked": null,
                            "count": 0,
                            "isDisabled": true,
                            "items": Array [
                              Object {
                                "onClick": [Function],
                                "title": "Select none (0)",
                              },
                              Object {
                                "onClick": [Function],
                                "title": "Select page (0)",
                              },
                            ],
                            "onSelect": [Function],
                          }
                        }
                        component="InventoryTable"
                        customFilters={
                          Object {
                            "filter": Object {
                              "system_profile": Object {},
                            },
                            "tags": undefined,
                          }
                        }
                        fallback={
                          <Bullseye>
                            <Spinner
                              size="xl"
                            />
                          </Bullseye>
                        }
                        getEntities={[Function]}
                        innerRef={null}
                        noDetail={true}
                        onLoad={[Function]}
                        showTags={true}
                        tableProps={
                          Object {
                            "canSelectAll": false,
                            "ouiaId": "systems-table",
                            "selectVariant": undefined,
                          }
                        }
                      />
                    }
                    appName="chrome"
                    bulkSelect={
                      Object {
                        "checked": null,
                        "count": 0,
                        "isDisabled": true,
                        "items": Array [
                          Object {
                            "onClick": [Function],
                            "title": "Select none (0)",
                          },
                          Object {
                            "onClick": [Function],
                            "title": "Select page (0)",
                          },
                        ],
                        "onSelect": [Function],
                      }
                    }
                    customFilters={
                      Object {
                        "filter": Object {
                          "system_profile": Object {},
                        },
                        "tags": undefined,
                      }
                    }
                    fallback={
                      <Bullseye>
                        <Spinner
                          size="xl"
                        />
                      </Bullseye>
                    }
                    getEntities={[Function]}
                    history={
                      Object {
                        "action": "POP",
                        "block": [Function],
                        "canGo": [Function],
                        "createHref": [Function],
                        "entries": Array [
                          Object {
                            "hash": "",
                            "pathname": "/",
                            "search": "",
                            "state": undefined,
                          },
                        ],
                        "go": [Function],
                        "goBack": [Function],
                        "goForward": [Function],
                        "index": 0,
                        "length": 1,
                        "listen": [Function],
                        "location": Object {
                          "hash": "",
                          "pathname": "/",
                          "search": "",
                          "state": undefined,
                        },
                        "push": [Function],
                        "replace": [Function],
                      }
                    }
                    innerRef={null}
                    module="./InventoryTable"
                    noDetail={true}
                    onLoad={[Function]}
                    scope="chrome"
                    showTags={true}
                    store={
                      Object {
                        "clearActions": [Function],
                        "dispatch": [Function],
                        "getActions": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                      }
                    }
                    tableProps={
                      Object {
                        "canSelectAll": false,
                        "ouiaId": "systems-table",
                        "selectVariant": undefined,
                      }
                    }
                  />
                </ForwardRef>
                <Bullseye>
                  <div
                    className="pf-l-bullseye"
                  >
                    <Spinner
                      size="xl"
                    >
                      <span
                        aria-valuetext="Loading..."
                        className="pf-c-spinner pf-m-xl"
                        role="progressbar"
                      >
                        <span
                          className="pf-c-spinner__clipper"
                        />
                        <span
                          className="pf-c-spinner__lead-ball"
                        />
                        <span
                          className="pf-c-spinner__tail-ball"
                        />
                      </span>
                    </Spinner>
                  </div>
                </Bullseye>
              </Suspense>
            </BaseInvTable>
          </ForwardRef>
        </SystemsTable>
      </Connect(SystemsTable)>
    </Provider>
  </Router>
</MemoryRouter>
`;

exports[`ConnectedSystemsTable should render radio correctly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(SystemsTable)
        hasInventoryReadPermissions={true}
      >
        <SystemsTable
          driftClearFilters={[Function]}
          hasInventoryReadPermissions={true}
          selectEntities={[Function]}
          selectHistoricProfiles={[Function]}
          selectedSystemIds={Array []}
          setSelectedSystemIds={[Function]}
          updateColumns={[Function]}
        >
          <Spinner
            size="lg"
          >
            <span
              aria-valuetext="Loading..."
              className="pf-c-spinner pf-m-lg"
              role="progressbar"
            >
              <span
                className="pf-c-spinner__clipper"
              />
              <span
                className="pf-c-spinner__lead-ball"
              />
              <span
                className="pf-c-spinner__tail-ball"
              />
            </span>
          </Spinner>
        </SystemsTable>
      </Connect(SystemsTable)>
    </Provider>
  </Router>
</MemoryRouter>
`;
